// @databricks/sql/dist/result/JsonResult.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Types_1 = require("../hive/Types");
const utils_1 = require("./utils");
class JsonResult {
    constructor(schema) {
        this.schema = (0, utils_1.getSchemaColumns)(schema);
    }
    getValue(data) {
        if (this.schema.length === 0 || !data) {
            return [];
        }
        return data.reduce((result, rowSet) => {
            const columns = rowSet.columns || [];
            const rows = this.getRows(columns, this.schema);
            return result.concat(rows);
        }, []);
    }
    getRows(columns, descriptors) {
        return descriptors.reduce((rows, descriptor) => this.getSchemaValues(descriptor, columns[descriptor.position - 1]).reduce((result, value, i) => {
            if (!result[i]) {
                result[i] = {};
            }
            const { columnName } = descriptor;
            result[i][columnName] = value;
            return result;
        }, rows), []);
    }
    getSchemaValues(descriptor, column) {
        var _a;
        const typeDescriptor = (_a = descriptor.typeDesc.types[0]) === null || _a === void 0 ? void 0 : _a.primitiveEntry;
        const columnValue = this.getColumnValue(column);
        if (!columnValue) {
            return [];
        }
        return columnValue.values.map((value, i) => {
            if (columnValue.nulls && this.isNull(columnValue.nulls, i)) {
                return null;
            }
            return (0, utils_1.convertThriftValue)(typeDescriptor, value);
        });
    }
    isNull(nulls, i) {
        const byte = nulls[Math.floor(i / 8)];
        const ofs = Math.pow(2, (i % 8));
        return (byte & ofs) !== 0;
    }
    getColumnValue(column) {
        if (!column) {
            return undefined;
        }
        return (column[Types_1.ColumnCode.binaryVal] ||
            column[Types_1.ColumnCode.boolVal] ||
            column[Types_1.ColumnCode.byteVal] ||
            column[Types_1.ColumnCode.doubleVal] ||
            column[Types_1.ColumnCode.i16Val] ||
            column[Types_1.ColumnCode.i32Val] ||
            column[Types_1.ColumnCode.i64Val] ||
            column[Types_1.ColumnCode.stringVal]);
    }
}
exports.default = JsonResult;
