// @databricks/sql/dist/DBSQLOperation/CompleteOperationHelper.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Status_1 = __importDefault(require("../dto/Status"));
class CompleteOperationHelper {
    constructor(driver, operationHandle, closeOperation) {
        this.closed = false;
        this.cancelled = false;
        this.driver = driver;
        this.operationHandle = operationHandle;
        this.closeOperation = closeOperation;
    }
    cancel() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.cancelled) {
                return Status_1.default.success();
            }
            const response = yield this.driver.cancelOperation({
                operationHandle: this.operationHandle,
            });
            Status_1.default.assert(response.status);
            this.cancelled = true;
            return new Status_1.default(response.status);
        });
    }
    close() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.closed && this.closeOperation) {
                Status_1.default.assert(this.closeOperation.status);
                this.closed = true;
            }
            if (this.closed) {
                return Status_1.default.success();
            }
            const response = yield this.driver.closeOperation({
                operationHandle: this.operationHandle,
            });
            Status_1.default.assert(response.status);
            this.closed = true;
            return new Status_1.default(response.status);
        });
    }
}
exports.default = CompleteOperationHelper;
